Sources/SwiftProtobuf/PathDecoder.swift

```diff
@@ -0,0 +1,426 @@
+  }
+
+  mutating func decodeRepeatedStringField(value: inout [String]) throws {
+    try setRepeatedValue(&value)
+  }
+
+  mutating func decodeSingularBytesField(value: inout Data) throws {
+    try setValue(&value, defaultValue: .init())
+  }
+
+  mutating func decodeSingularBytesField(value: inout Data?) throws {
+    try setValue(&value, defaultValue: nil)
+  }
+
+  mutating func decodeRepeatedBytesField(value: inout [Data]) throws {
+    try setRepeatedValue(&value)
+  }
+
+  mutating func decodeSingularEnumField<E>(
+    value: inout E
+  ) throws where E : Enum, E.RawValue == Int {
+    try setValue(&value, defaultValue: .init())
+  }
+
+  mutating func decodeSingularEnumField<E>(
+    value: inout E?
+  ) throws where E : Enum, E.RawValue == Int {
+    try setValue(&value, defaultValue: nil)
+  }
+
+  mutating func decodeRepeatedEnumField<E>(
+    value: inout [E]
+  ) throws where E : Enum, E.RawValue == Int {
+    try setRepeatedValue(&value)
+  }
+
+  mutating func decodeSingularMessageField<M>(
+    value: inout M?
+  ) throws where M : Message {
+    if nextPath.isEmpty {
+      try setValue(&value, defaultValue: nil)
+      return
+    }
+    var decoder = try PathDecoder<M>(
+        path: nextPath,
+        value: self.value,
+        mergeOption: mergeOption
+    )
+    if value == nil {
+      value = .init()
+    }
+    try value?.decodeMessage(decoder: &decoder)
+  }
+
+  mutating func decodeRepeatedMessageField<M>(
+    value: inout [M]
+  ) throws where M : Message {
+    try setRepeatedValue(&value)
+  }
+
+  mutating func decodeSingularGroupField<G>(
+    value: inout G?
+  ) throws where G : Message {
+    try setValue(&value, defaultValue: nil)
+  }
+
+  mutating func decodeRepeatedGroupField<G>(
+    value: inout [G]
+  ) throws where G : Message {
+    try setRepeatedValue(&value)
+  }
+
+  mutating func decodeMapField<KeyType, ValueType>(
+    fieldType: _ProtobufMap<KeyType, ValueType>.Type,
+    value: inout _ProtobufMap<KeyType, ValueType>.BaseType
+  ) throws where KeyType : MapKeyType, ValueType : MapValueType {
+    try setMapValue(&value)
+  }
+
+  mutating func decodeMapField<KeyType, ValueType>(
+    fieldType: _ProtobufEnumMap<KeyType, ValueType>.Type,
+    value: inout _ProtobufEnumMap<KeyType, ValueType>.BaseType
+  ) throws where KeyType : MapKeyType, ValueType : Enum, ValueType.RawValue == Int {
+    try setMapValue(&value)
+  }
+
+  mutating func decodeMapField<KeyType, ValueType>(
+    fieldType: _ProtobufMessageMap<KeyType, ValueType>.Type,
+    value: inout _ProtobufMessageMap<KeyType, ValueType>.BaseType
+  ) throws where KeyType : MapKeyType, ValueType : Hashable, ValueType : Message {
+    try setMapValue(&value)
+  }
+
+  mutating func decodeExtensionField(
+    values: inout ExtensionFieldValueSet,
+    messageType: any Message.Type,
+    fieldNumber: Int
+  ) throws {
+    preconditionFailure(
+      "Path decoder should never decode an extension field"
```

Comment:
Prefix this with "Internal Error:" just so if it ever did show up it might be a little more clear for folks.